src/libpgmoneta/message.c

```diff
@@ -2370,134 +2370,117 @@ pgmoneta_receive_extra_files(SSL* ssl, int socket, char* username, char* source_
-            }
+            dest_dir = (char*)malloc((strlen(target_dir) + 1) * sizeof(char));
+            memset(dest_dir, 0, strlen(target_dir) + 1);
+            memcpy(dest_dir, target_dir, strlen(target_dir));
+         }
 
-            pgmoneta_mkdir(dest_dir);
+         pgmoneta_mkdir(dest_dir);
 
-            dest_path = (char*)malloc((strlen(dest_dir) + strlen(file_name) + 1) * sizeof(char));
-            snprintf(dest_path, strlen(dest_dir) + strlen(file_name) + 1, "%s%s", dest_dir, file_name);
+         dest_path = (char*)malloc((strlen(dest_dir) + strlen(file_name) + 1) * sizeof(char));
+         snprintf(dest_path, strlen(dest_dir) + strlen(file_name) + 1, "%s%s", dest_dir, file_name);
 
-            pgmoneta_ext_get_file(ssl, socket, paths[j], &qr);
-            if (qr != NULL && qr->tuples != NULL)
+         pgmoneta_ext_get_file(ssl, socket, paths[j], &qr);
+         if (qr != NULL && qr->tuples != NULL)
+         {
+            struct tuple* current_tuple = qr->tuples;
+            while (current_tuple != NULL)
             {
-               struct tuple* current_tuple = qr->tuples;
-               while (current_tuple != NULL)
+               if (current_tuple->data != NULL && current_tuple->data[0] != NULL)
                {
-                  if (current_tuple->data != NULL && current_tuple->data[0] != NULL)
-                  {
-                     int decoded_len;
-                     unsigned char* decoded_data;
+                  int decoded_len;
+                  unsigned char* decoded_data;
 
-                     decoded_data = decode_base64(current_tuple->data[0], &decoded_len);
-                     if (decoded_data != NULL)
+                  decoded_data = decode_base64(current_tuple->data[0], &decoded_len);
+                  if (decoded_data != NULL)
+                  {
+                     FILE* file = fopen(dest_path, "wb");
+                     if (file != NULL)
                      {
-                        FILE* file = fopen(dest_path, "wb");
-                        if (file != NULL)
-                        {
-                           fwrite(decoded_data, 1, decoded_len, file);
-                           fclose(file);
-                        }
-                        else
-                        {
-                           pgmoneta_log_error("Retrieving extra files: Could not open file \"%s\" for writing", dest_path);
-                           free(dest_dir);
-                           free(dest_path);
-                           goto error;
-                        }
-                        free(decoded_data);
+                        fwrite(decoded_data, 1, decoded_len, file);
+                        fclose(file);
                      }
+                     else
+                     {
+                        pgmoneta_log_error("Retrieving extra files: Could not open file \"%s\" for writing", dest_path);
+                        free(dest_dir);
```

Comment:
Probably define dest_dir and dest_path out of the scope so that you can handle free in the error section.